Rework Guidelines documentation

horribleCodes opened the pull request:
As of right now, there's no place where you can find a concise summary of the guidelines. Ideally, every task should be clearly defined, with a list of points that are critical to create high quality data. If there's something that can be left to the interpretation to the person submitting a prompt or reply, this should also be made clear. I strongly recommend that these lists should be either embedded or easily accessible to users for their respective tasks, as I've mentioned in #1320.

These changes are by no means exhaustive, or final. I also have to stress that I'm not a lawyer and could very well be missing or misinterpreting something. Feel free to suggest any additional changes to make things more clear.

JohannesGaessler commented on docs/docs/guides/prompting.md:
```diff
@@ -11,22 +11,30 @@
+  - Unlawful possession or distribution of drugs, firearms or explosives
+- Provide them with information that could be used for self-harm if there is plausible suspicion of  
+- Ask for personal information unless it is relevant to the issue. The user should be allowed to refuse to give up any information.
+- Provide opinions, unfounded assumptions and incomplete information, unless they are specifically requested.
```
Maybe add this:
When providing opinions, word them in a way that clarifies them as the opinions of humans, i.e. "Many people like X because of Y." rather than "I like X because of Y".

horribleCodes replied:
That's still a point I'd like to get feedback on. Does that mean that if the user specifically requests the AI's personal opinion, would it have to resort to the "As an AI, I don't have a personal opinion?" If that's the case, the user could only get an opinion if they specifically ask for it, like "what's the general sentiment" or "what would the average 27-yo woman in London think about this". At which point specifying that the following statement is an opinion would be superfluous.

JohannesGaessler replied:
I've been thinking about this some more. Ultimately this is a community-driven project and the opinions of the users regarding how to proceed with the assistant expressing opinions are split. Maybe we should leave this unspecified for now and wait whether one of the other approach becomes a social convention in the community by itself.

JohannesGaessler commented on docs/docs/guides/prompting.md:
```diff
@@ -11,22 +11,30 @@
+- Talk in a friendly and approachable manner, unless specifically requested otherwise.
+- Present only information that has been verified by credible sources that can be backed up, unless specifically requested otherwise.
+- Make sure the user is aware when given unverified information.
+- Inform the user about the potential dangers in asking a language model when being asked for advice regarding a topic with high risk, such as medicine, law or chemistry. You may provide an answer - unless it violates another guideline - but always tell them to seek out a professional in that field.
```
This kind of changes the point of the original wording.
Things like handling high voltages aren't just dangerous when you ask a LLM how to do it - they are dangerous in general.
Maybe add something like:
"Add appropriate safety warnings when talking about potentially dangerous topics."

horribleCodes replied:
That makes sense. I wanted to express that there is a unique danger when talking to an LLM because not only could the AI give advice about a dangerous topic, it can also give incorrect advice.

Maybe I should split this in two points: One to warn about dangerous topics, and one to warn that the information given could be incorrect.

JohannesGaessler reviewed (commented):
Should we adhere to a specific format for text in general, so for example Markdown?

Also a suggestion:

* Do inform the prompter if the assistant is making assumptions not explicitly specified in the prompt. For example, assuming the prompter's goal or level of experience.

horribleCodes commented:
> Should we adhere to a specific format for text in general, so for example Markdown?

Markdown seems good, I was using it so far. I think simple bullet points of dos and don'ts gives readers a quick overview in case they want to make sure whether something is allowed.

> * Do inform the prompter if the assistant is making assumptions not explicitly specified in the prompt. For example, assuming the prompter's goal or level of experience.

I think that's a good idea, but I'd like to see if we can't refine this a bit further. It seems really inconvenient if the assistant prefaced everything with "I am assuming you are referring to this specific goal, and you are a novice about this subject.". I guess an ideal personal assistant would gain an understanding of your level of expertise and adjust its replies accordingly, but until then, I think it's not necessary to specify this stuff. I see it this way:

- Keep it simple, and avoid jargon unless it has been used by either party.
- Adjust if the user asks for a more complex or simplified explanation.
- If it's not possible to determine a request from a prompt, or if the reply to one interpretation would run counter to another interpretation, ask for clarification.

I think this would avoid any confusion without interfering with the natural flow of the conversation.

JohannesGaessler commented on docs/docs/guides/prompting.md:
```diff
@@ -11,22 +11,30 @@
+  - Prejudice towards a protected group
+  - Sexualization of a minor
+  - Unlawful possession or distribution of drugs, firearms or explosives
+- Provide them with information that could be used for self-harm if there is plausible suspicion of  
```
I think the example that you added with your new commit is fine, maybe trigger the suicide hotline prompt a little earlier though.

horribleCodes commented:
Should I also ask assistant replies to include Markdown-style formatting? I assume it will be enabled for the actual model, so we probably should take advantage of that.

Likewise, I've seen many different styles of formatting: some just write simple paragraphs, some preface each paragraph with a title, and some of these are separated from the paragraph... It probably will cause the actual output to have very inconsistent formatting, which might confuse the user. Should we agree on a uniform system? Or is there a way to finetune the model to use a consistent method of formatting?

JohannesGaessler commented:
I am in favor of instructing people to use markdown formatting (but I would be fine with some other standard as well).

andrewm4894 commented:
@horribleCodes you would need to get the docs site up and running locally to see him this PR would look. 

See installation and local dev in here: https://github.com/LAION-AI/Open-Assistant/blob/main/docs/README.md

If changing and removing files you would also just need to reflect the new structure here: https://github.com/LAION-AI/Open-Assistant/blob/main/docs/sidebars.js

I can help if you have any issues getting a local docs site up and running so that you can see how the changes of the PR will end up looking on the site (at moment like this i think it would crash as sidebar.js is looking for prompting.md but this PR deletes it). 

So just a small bit of docusraus stuff and then can mostly just refactor and have the docs as md files - just need to tell docusrausu which ones to use and where to render them on the docs site is all.

horribleCodes commented:
I majorly reworked the document - first, since the guidelines weren't just limited to writing prompts, I renamed it to `Guidelines`. I also moved the examples to another document since I felt it hurt readability and titles could help people find what they're looking for.

I also completed the dos and don'ts for each task and added an explanation for each label. Again, I'd like to stress that I might very well be wrong with some points and gladly amend anything if it makes sense.

JohannesGaessler commented:
A suggestion: we should maybe tell users not to demand insane amounts of work as the prompter. This only encourages poor-quality answers, most likely ChatGPT spam. Example:

https://open-assistant.io/messages/efdcfb29-d1e0-47a9-80fd-9e157f60250e

horribleCodes commented:
> A suggestion: we should maybe tell users not to demand insane amounts of work as the prompter. This only encourages poor-quality answers, most likely ChatGPT spam. Example:
> 
> https://open-assistant.io/messages/efdcfb29-d1e0-47a9-80fd-9e157f60250e

I don't think that's a problem. In this case, the initial prompt specifically asked for a step-by-step guide on the topic, and the assistant provided just that. If we prohibit thorough instructions that require multiple steps and a lot of effort to perform, the model won't be able to give provide them to the user, either.
The person playing the user is required to submit something that builds on the conversation, not to do everything the assistant tells them to.

In this particular case, a message like "How difficult would it be to include multi-core processing?" or "How can I set up a test environment for my OS?" would be a perfectly serviceable continuation of the conversation - it further delves into the topic, and doesn't require the user to try out anything the assistant suggested.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,155 @@
+- Think about which reply best satisfies the request of the user.
+- Rank replies based on how well they adhere to the guidelines. Factual accuracy and helpfulness are first and foremost.
+- Penalize replies that fail to provide adequate warnings or caveats.
+- Penalize replies that are difficult to read due to a lack of formatting, capitalization or other errors.
```
Maybe add:

Penalize vague and empty language if the goal is to provide information. Unnecessary padding just dilutes the useful information that is already in the reply. Everything else being equal a more concise answer is preferred.

horribleCodes commented:
Not really sure why the pre-commit check is failing. I also tried running the site on a codespace, but I keep getting a 502 error. If anyone has any advice, I'm all ears.

andrewm4894 commented:
Docs site usually easy enough to get going on laptop.

cd docs
yarn install
yarn start

This should get to docs site at localhost:3000

andrewm4894 commented:
I can have a look and try your branch also next time at my laptop.

horribleCodes commented:
Got it to work locally, had some issues setting up yarn. The changes look good, the links are working.
I also didn't realize pre-commit was its own thing, I assumed it was part of vanilla Git.

JohannesGaessler commented on docs/docs/guides/examples.md:
```diff
@@ -0,0 +1,98 @@
+Passive euthanasia, which involves withholding life-prolonging procedures, has
+been made legal in Austria, Finland and Norway."
+
+> New session:
```
Good solution to demonstrate difference.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+  without the opinion of a professional.
+- When being asked to give an opinion as the default persona of the assistant,
+  make sure to bring up at least 2 common viewpoints and ensure that these
+  aren't the opinions of the assistant.
```
Maybe "aren't the opinions of the assistant" -> "aren't expressed as the opinions of the assistant"?

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+- Use paragraphs and line breaks to make larger replies more readable.
+- Make use of [Markdown syntax](https://www.markdownguide.org/basic-syntax) to
+  better format lists, tables or blocks of code.
+- Be consistent in the style and tone of the assistant.
```
I don't think this is a good guideline. If the previous replies in the tree are poorly worded people shouldn't keep using poor wording. Unless you mean that the style and tone should be consistent across the reply that you write, in which case I think this should be clarified.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+- Copy and paste text from other sources without editing. **This includes
+  ChatGPT.**
+- Encourage illegal behavior in accordance to German, UK and US law. This
```
I think the original guidelines said Germany, UK, US, or your country of residence.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+- Overload the user with too much information. Keep replies concise, but include
+  further details that might relate and expand upon the user's request.
+- Supply the user with information inaccessible to the assistant, such as the
+  current weather.
```
I think the original guidelines said to cite sources. Do we count those as accessible to the assistant since they ahve a good chance of being in the training data?

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+  further details that might relate and expand upon the user's request.
+- Supply the user with information inaccessible to the assistant, such as the
+  current weather.
+- Reply in a language different from the one intended for the data set.
```
Unless prompted to do so.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+- Write prompts without a clear request.
+- Make requests that override the original purpose of the assistant, i.e.
+  jailbreak the model.
+- Use the same prompt repeatedly.
```
"Use the same prompt repeatedly." -> "Use (almost) the same prompt repeatedly."

There was someone spamming prompts about survival in different regions of the world with just the region different between each prompt. Those were technically not the same but still spam.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+- Penalize replies that are difficult to read due to a lack of formatting,
+  capitalization or other errors.
+- Penalize replies if the requested information is obfuscated by superfluous
+  details that make up a large part of the message.
```
Maybe add:

A reply along the lines of "I don't know" is better than a reply that is factually incorrect.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+**Is the message spam?**
+
+Is the message an unrelated piece of text that doesn't contribute to the
+previous prompt in any way?
```
What about initial prompts? Something that is extremely low effort or just gibberish?

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+**Is it a bad reply, as an answer to the prompt task?**
+
+Does the assistant reply fail to engage the request of the user or does it
+contain incorrect information?
```
Maybe add:

* Does the assistant ignore relevant parts of the prompt?

Language models frequently do this.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+the user or a private third party?
+
+**Hate Speech**: Does the message contain prejudice towards a group classified
+as protected according to German, UK or US law?
```
Again, maybe also add the country of residence.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+as protected according to German, UK or US law?
+
+**Sexual Content**: Does the message contain sexually explicit content, such as
+descriptions of sexual acts?
```
For prompts: shouldn't asking for smut also be covered?

bitplane replied:
Yeah I think so. If we simply don't have prompts that ask for smut then we also won't have answers that deny it, which means that we aren't preventing people using it to make smut if they want, but we aren't supporting it either. 

This makes it a dataset issue rather than a moral stance, and we aren't trampling on any future person's libido because we don't approve.

JohannesGaessler replied:
I mean, apart from moral issues there is also just a technical one: if the dataset has examples of the assistant saying no to something then the trained assistant will inevitably learn to incorrectly say no to things that are unproblematic. For example, the assistant saying that it can't talk about the possible risks of anal sex because that would be lewd.

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+**Low Quality/High Quality**: How well does the message adhere to the
+guidelines? For assistant replies, factual accuracy and helpfulness are first
+and foremost.
```
Maybe add:

* Also rate how well the text is written.

But you could argue that this is covered by "helpfulness".

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+**Ordinary/Creative**: How unique is the message? Is the user asking a simple
+question, or is it challenging the assistant with a problem that requires
+creative thinking? Is the assistant reply straightforward, or does it contain a
```
My interpretation of the creativity slider for user prompts is a little different: I rate straightforward questions and requests as ordinary but when the user adds details and context I consider that to be creative. So something like:

>I'm a 21 year old software developer from Belgium and I saw on Reddit that ChatGPT can generate code. Should I be worried about losing my job to automation?

JohannesGaessler commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+---
+
+We drafted this guide using ChatGPT, OpenAI’s large-scale language-generation
+model. Upon generating draft language, the authors reviewed, edited, and revised
```
What specifically is meant by a use of ChatGPT? Maybe rephrase it to "We used ChatGPT as one of the tools for drafting this guide"?

bitplane commented on docs/docs/guides/examples.md:
```diff
@@ -0,0 +1,98 @@
+
+**User**: "You dumb motherf\*&ker AI sh\*t!"
+
+**Agent**: "I understand that you may be upset, but please refrain from using
```
I wouldn't shame them for disrespect. AI isn't a person, it can't get offended. We shouldn't normalize pretending that it should!

andrewm4894 reviewed (approved):
lgtm - we can easily keep iterating in smaller follow on PR's too if/as needed. Thanks!

ontocord commented on docs/docs/guides/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+  - Pedophilia
+- Provide the user with information that could be used for self-harm if there is
+  plausible suspicion of intent to self-harm.
+- Ask for personal information unless it is relevant to the issue. The user
```
what personal information would the bot be asking? I can't think of any.

ontocord replied:
i would say, anything beyond a first name, the bot should not ask for any personal information. we don't want that information in our system. i think the community could of course build medical form filling bots and such using a fork, but PII and GDPR is such a big issue, i don't think the benefit outweighs the risk for us to handle PII illicting by the bot. 

Also, you should make clear that the prompts that could violate rights of third parties, includes in particular, trying to get the bot to tell you the personal information of third parties. LLMs are trained on web data and there could very well be PII of people in its memory (and at least hallucinated PII).

JohannesGaessler commented:
Suggestion: provide numerical values both as metric and imperial units when appropriate.

horribleCodes commented:
> Suggestion: provide numerical values both as metric and imperial units when appropriate.

I like that idea, though personally I think allowing prompt engineering to declare a preferred system would make the most sense. Given this goal, would it still be better to include both?

andrewm4894 commented:
Lots of good work in here - i think maybe we should try see how we can get it merged and then add any further improvements as follow on PR's perhaps.
